metricbeat/module/elasticsearch/pending_tasks/data_test.go

```diff
@@ -80,24 +66,18 @@ func TestInvalidJsonForRequiredFieldShouldThrowError(t *testing.T) {
 	content, err := ioutil.ReadFile(file)
 	assert.NoError(t, err)
 
-	_, errs := eventsMapping(content)
-
-	errors, ok := errs.(*s.Errors)
-	if ok {
-		assert.True(t, errors.HasRequiredErrors(), "mapping error: %s", errors)
-		assert.EqualError(t, errors, "Required fields are missing: ,source")
```

Comment:
@ruflin this is the only place where I have seen that we'd make use of `ApplyResult.NotFound`, to check if `source` is in the list of missing fields, but to take advantage of this we'd have to complicate the `eventsMapping` function to return the result and not only the event and the errors. Also, I think that here it'd be enough with testing that it returns an error for the case of having a missing field, going further looks like if we were testing here the behaviour of `schema.Apply`, that is already being tested in the `schema` package.

Given that, I wonder again if it worths it to have the `ApplyResult` object. From the perspective of a consumer of the API, it is cleaner to just have `event, err := schema.Apply(data)`, with an event and an error that can be used directly. I think it is also more valuable to be able to give directly the semantics of the use case, as in `event.MetricSetFields, err = schema.Apply(index)` or `nodeData, _ := schemaXpack.Apply(node)` instead of obtaining the result and then assigning `result.Event` to a meaningful variable. From our perspective, if we want to test more particular cases of missing fields, I think that a better place would be the `schema` package.

I'd go back to the approach in #7167, but with the additional improvements in error handling added here too.

Reply:
I can see that from a user perspective it is much more convenient if the event can be assigned directly. It's also the reason why I started to think about 3 arguments.

For testing for the missing fields which are optional I think this is something specific to metricsets and not necessarly internal testing of Apply. The reason it only exists in one place so far I think is more that it was too complicated to do it. When we start to add more versions and have very subtle differences between each version I think tests for missing fields become more common. 

I assume moving forward we will make optional fields the default as we discussed in the past and not required. In this scenario most fields will be optional. So I our testing will potentially more look like: max 3 missing fields and we get an error if there are more.

Reply:
I have changed the approach again, to `event, err := schema.Apply(data, options...)`, `options` is optional, and can be used to select the mode of operation of `Apply` (so for example we can make all fields optional by default). Or to access unfiltered errors, so additional tests can be done on the missing fields for example. The test that was checking the missing fields has been refactored to this approach.

I have updated the description of the PR accordingly.